### ISSUE TITLE
Manual / External Chunk / Data Lifecycle API

### ISSUE THREAD

**Issue by alamb** [influxdata/influxdb_iox#769]
__Use case:__

As explained in the "Data Lifecycle in IOx" [blog](https://www.influxdata.com/blog/how-influxdb-iox-manages-the-data-lifecycle-of-time-series-data/) and [talk](https://www.youtube.com/watch?v=KwdPifHC1Gc), data moves through IOx and eventually is stored in Chunks.

Chunks are initially created in the mutable buffer, and then eventually "closed" and migrated to either the read buffer and/or Parquet files on object storage

Paul envisions both a "self preservation" mode where basic data lifecycle is dictated in response to a simple policy, but also the ability for an external system to control this motion in a more fine grained manner.

The lifecycle of chunks looks roughly like:
```
... -> Mutable Buffer ---chunk---> Read Buffer
                      \--chunk---> Parquet Files
```
(and in the future there will be a line from parquet files --> read buffer)

Part of this motion have been implemented in the [Db](https://github.com/influxdata/influxdb_iox/blob/main/server/src/db.rs) structure

__Proposal:__
This ticket proposes some external HTTP API routes to control the chunk lifecycles. This ticket likely needs to be broken down into smaller tasks

# Partition CRUD
`GET /iox/api/v1/databases/:name/partitions`: Lists partitions known to the database
`GET /iox/api/v1/databases/:name/partitions/:partition_key`: get a partition with the specified key
`DELETE /iox/api/v1/databases/:name/partitions/:partition_key`: delete a partition (and all of its chunks)

## Partition Entity:

```
[{
  partition_key: "2012-01-01",
  # probably other useful fields here
},
{
  partition_key: "2012-02-01",
  # probably other useful fields here
}
...
]
```

# Chunk CRUD
`GET /iox/api/v1/databases/:name/partitions/:partition_key/chunks`: Lists chunks available in database
`GET /iox/api/v1/databases/:name/partitions/:partition_key/chunks/:chunk_id`: get just specific chunk
`PUT /iox/api/v1/databases/:name/partitions/:partition_key/chunks/:chunk_id`: request an update to a specific chunk
`DELETE /iox/api/v1/databases/:name/partitions/:partition_key/chunks/:chunk_id`: delete a chunk

## Chunk Entity

Question: do we like the `object_store_path` representation?

```
[{
   chunk_id: 42,
   open: false,
   location: ["mutable_buffer", "read_buffer", "object_store"],
   object_store_path: "path/to/parquet/file/in/object/store"
}
..
{
   chunk_id: 44,
   open: false,
   location: ["mutable_buffer", "read_buffer"],
}
...
{
   chunk_id: 45,
   open: true,
   location: ["mutable_buffer"],
}]
```

### Step 1. Get the Original Chunk
```
GET /iox/api/v1/databases/my_db/partitions/2021-01-01/chunks/45
```

This returns:
```
{
   chunk_id: 45,
   open: true,
   location: ["mutable_buffer"],
}
```

### Step 2: Update chunk

```
POST /iox/api/v1/databases/my_db/partitions/2021-01-01/chunks/45
```
with the following Post body:
```
{
   chunk_id: 45,
   open: false,
   location: ["mutable_buffer", "read_buffer"],
}
```
Response will be HTTP 202 with a `Location: /iox/api/v1/jobs/XXX` header

### Step 3: Get job status:

```
GET /iox/api/v1/jobs/1234
```

Response is 200

```
{
    "status": "waiting",
    "issuedAt": "2020-10-03T09:34:24",
    "request": {
        chunk_id: 45,
        ...
    }
    "links": [{
        "rel": "cancel",
        "method": "delete",
        "href": "/iox/api/v1/jobs/1234"
    }]
}
```

In the background, the server will then:
* call [rollover_partition](https://github.com/influxdata/influxdb_iox/blob/main/server/src/db.rs#L116)
* call [load_chunk_to_read_buffer](https://github.com/influxdata/influxdb_iox/blob/main/server/src/db.rs#L198)

### Step 4: Job is done
When the server has completed the request, it will return a pointer back to the chunk
```
GET /iox/api/v1/jobs/1234
```

Returns a HTTP 301 (see other) with a `Location: /iox/api/v1/databases/my_db/partitions/2021-01-01/chunks/45` HTTP header

For some additional examples of this style REST API, see: https://www.mscharhag.com/api-design/rest-asynchronous-operations

# Open Questions:
* How do we handle conflicting requests (e.g. one request to load the chunk to read_bufer and one to move it to object_store) come in semi-concurrently?

**Comment by alamb** [comment]
@pauldix  what do you think?

**Comment by pauldix** [comment]
I think it might be a little cleaner to separate the actions that move or delete chunks or partitions to sub-areas below `.../partitions`. There are limited things you can do with a chunk so I think it would be easier to use if those are called out as explicit POST operations.

For example, do you want to delete the partition from object storage, the mutable buffer, the read buffer, or some combination thereof? Also, a chunk should only ever be able to be in either the mutable buffer or the read buffer, but never both.

Should the summaries be included in this response? I realize that the current `PartitionSummary` doesn't have `Chunk` represented, but maybe it should? Or maybe these should be separate concerns, but ultimately, I think this is what is going to be persisted in the catalog and used for query planning. So I'd like to give that same response back to the user.

Or maybe optionally include the summaries, or have one endpoint that includes them and one that doesn't. For reference, we'd expect users to have potentially hundreds or even low thousands of individual tables, which would end up showing up in the summaries.

The current code only allows for a single open chunk in the mutable buffer per partition. Are you thinking this might change? There was one point where we were talking about having a third state for chunks in the mutable buffer where they're open for writes if those writes fall within a certain range, otherwise those writes would be put into new chunks. But I think it might be easier to just avoid that and suggest that partition sizing be set such that you commonly don't close a chunk until that partition is no longer receiving writes.

I'd also like to see estimated (or actual) size of chunks in the response. Although maybe that is best included with the summary info?

Some ideas here:

Update the `partition_metadata.rs` to have `PartitionSummary` like: 

```rust
pub struct PartitionSummary {
    pub key: String,
    pub chunks: Chunk,
}

pub struct Chunk {
	pub id: u32,
	pub tables: Vec<TableSummary>,
	pub size: usize,
	pub in_storage: bool,
	// these two feel weird since they should be mutually exclusive,
	// but we may change this later, so might be best to have this
	pub in_mutable_buffer: bool,
	pub in_read_buffer: bool,
}
```

I would probably have storage locations via a separate API, not in the summaries. Those are underlying details that ideally people won't be touching or working with.

```
# show partition summaries, which includes what chunks exist and their locations
GET /iox/api/v1/databases/:name/partition_summaries
# show the partition summary for a specific partition
GET /iox/api/v1/databases/:name/partition_summaries/:key

# drop a partition from the mutable buffer
DELETE /iox/api/v1/databases/:name/mutable_buffer/partitions/:key
# drop a chunk from mb
DELETE /iox/api/v1/databases/:name/mutable_buffer/partitions/:key/chunks/:id

# drop a partition from the read buffer
DELETE /iox/api/v1/databases/:name/read_buffer/partitions/:key
# drop a chunk from mb
DELETE /iox/api/v1/databases/:name/read_buffer/partitions/:key/chunks/:id

# delete a partition from object storage
DELETE /iox/api/v1/databases/:name/storage/partitions/:key
# delete a chunk from object storage
DELETE /iox/api/v1/databases/:name/storage/partitions/:key/chunks/:id

# close the open chunk, this happens immediately so should return 200, with the chunk ID that just got rolled
POST /iox/api/v1/databases/:name/mutable_buffer/partitions/:key/rollover

# start a job to persist the mutable buffer chunk to storage
POST /iox/api/v1/databases/:name/mutable_buffer/persist_jobs
# example body: {partition_key: "2021-02-08T19:00", chunk_id: 2}

# start a job to move the mutable buffer chunk over to the read buffer
POST /iox/api/v1/databases/:name/mutable_buffer/convert_jobs
# example body: {partition_key: "2021-02-08T19:00, chunk_id: 2"}

# load a partition from object storage, into the read buffer
POST /iox/api/v1/databases/:name/partition_load_jobs
# example body: {writer_id: 2, key: "2021-02-08T19:00"}

# get all jobs running for the db, or filter by a specific type
GET /iox/api/v1/databases/:name/jobs
PARAMS: type=persist|convert|load

# cancel a job
# I'd use a guid for the ID so we don't have to worry about server restarts reusing ids
DELETE /iox/api/v1/databases/:name/jobs/:id
```

I like what you had in there on the redirect for jobs.

I'd hold off on bothering with GET requests for individual partition chunks. As for getting the data, it seems like we'd have an API where the client would make requests for individual tables within a chunk/partition if they're getting in bulk, but it only makes sense for the stuff that is persisted to object storage, unless we construct the Parquet data on the fly, which we can certainly do. I'd like to think of the bulk data API separately from the metadata and data lifecycle API.

## Questions

1. To get into object storage, must a Chunk come from the MutableBuffer?

**Comment by tustvold** [comment]
Linking #655 for the jobs API

**Comment by alamb** [comment]
@pauldix I think the outline you sketched above looks good. 

> Or maybe optionally include the summaries

How about adding a query parameter  like `?include_summaries=true` so that the Partition/Chunk summaries will be included

> To get into object storage, must a Chunk come from the MutableBuffer?

TLDR; NO. I think it will be fairly easy to persist Chunks from MutableBuffer and ReadBuffer to object storage (they just need to make RecordBatches, which they can both do) so there is no implementation reason for this restriction

## On chunk / partition ids 

Something I heard from UI devs / app devs when working with APIs was that they were used to a *single* ID for the objects that are exposed and that multi-part keys was a pain to work with in the existing frameworks. So in addition to the API routes you specified above, I think devs / tools would want routes like

```
GET /iox/api/v1/partitions/:id
GET /iox/api/v1/partitions
DELETE /iox/api/v1/partitions/:id
```

And 
```
GET /iox/api/v1/chunks/:id
GET /iox/api/v1/chunks
DELETE /iox/api/v1/chunks/:id
```

The difference being that they need only `id` and not also `databse_name` and `partition_key`. This implies that we would need to somehow make `ids` unique across databases

I think the existing InfluxDB cloud API has this pattern for its object:
https://docs.influxdata.com/influxdb/cloud/api/#operation/GetSourcesIDBuckets

```
/api/v2/sources/{sourceID}/buckets
```

But then also lets you list all buckets with:
```
/api/v2/buckets
```

**Comment by pauldix** [comment]
I actually have the opposite experience working with APIs when resources are nested. Basically, if a resource "belongs to" another, meaning it has a reference back to some other resource, I would expect the ID of the parent to show up in the path and I would also expect that parent ID to always be returned in a response to get that child resource.

I wouldn't mirror the v2 API. As it has some things that are problematic. The request to list buckets, for example, requires an organization ID, it just doesn't happen to show up in the path, which I think is confusing.

Having things in the path makes it more obvious to the user what the relationships  between resources is, which I think is a win. Of course, I recognize that all of this is aesthetic so it really can be argued either way.

**Comment by alamb** [comment]
> I recognize that all of this is aesthetic so it really can be argued either way.

Yeah -- I am just parroting back my experience in the past where @choochburger (👋 ) showed me how React was setup to work super easily when each object had a global ID (ala the V2 API) but managing multi-part keys was a pain.

I don't have a strong preference either way, I was just trying to provide some past history (e.g. if we pick chunk ids like `0`, `1, `2`, etc that aren't unique we can't then retrofit guids on them without pain).... That might have happened to me in the past -- lol
